addons/mail/static/src/models/message/message.js

```diff
@@ -106,17 +114,47 @@ function factory(dependencies) {
                     ];
                 }
             }
-            if ('channel_ids' in data && data.channel_ids) {
-                // AKU FIXME: side-effect of calling convert...
-                const channelList = [];
-                for (const channelId of data.channel_ids) {
-                    const channel = this.env.models['mail.thread'].insert({
-                        id: channelId,
-                        model: 'mail.channel',
-                    });
-                    channelList.push(channel);
+            if ('moderation_status' in data) {
+                data2.moderation_status = data.moderation_status;
+                const moderationMailbox = this.env.messaging.moderation;
+                const originThread =
+                    data.res_id &&
+                    data.model &&
+                    this.env.models['mail.thread'].find(
+                        t => t.id === data.res_id && t.model === data.model
+                    );
+
+                if (data.moderation_status === 'pending_moderation') {
+                    // Server will return the message without any channels as it's
+                    // in pending_moderation, so we need to add it to its moderation box
+                    // AND in the origin channel manually for the current partner if the message is
+                    // moderated by him
+                    if (moderationMailbox) {
+                        if (originThread && originThread.isModeratedByCurrentPartner) {
+                            data2.threadCaches.push(
+                                ['link', moderationMailbox.mainCache],
+                                ['link', originThread.mainCache],
+                            );
```

Comment:
probably depending on current rework of threadCache by alex, but I really feel like this whole logic should be done generic at model level (compute/onchange of moderation_status ) and not at convertData

Reply:
> probably depending on current rework of threadCache by alex, but I really feel like this whole logic should be done generic at model level (compute/onchange of moderation_status ) and not at convertData

Indeed, with redesign of links between messages and threads, all this will be automated. This includes no longer requiring to manually link messages and caches, and linking manually links to threads is rare or super flow specific.

My branch is still work in progress... but the changes are significant and nice enough that I would wait for it before re-considering this PR, because it might just likely fix this PR at the same time, while making all messages/threads/caches links super robust and mostly automated (while also solving the "holes in message list" issues)

Reply:
Unless you are ready to merge today, I think we should merge this PR first as it is pretty much ready now (?) and already improves the situation (and introduces test to validate the effects of your PR too).